ctk/idp/src/main/kotlin/org/codice/compliance/web/slo/endpoint/ServiceProviderEndpointVerification.kt

```diff
@@ -0,0 +1,151 @@
+import org.codice.security.saml.SamlProtocol
+import org.codice.security.sign.Encoder
+import org.joda.time.DateTime
+import org.opensaml.saml.common.SAMLVersion
+import org.opensaml.saml.saml2.core.LogoutRequest
+import org.opensaml.saml.saml2.core.impl.IssuerBuilder
+import org.opensaml.saml.saml2.core.impl.LogoutResponseBuilder
+import org.opensaml.saml.saml2.core.impl.StatusBuilder
+import org.opensaml.saml.saml2.core.impl.StatusCodeBuilder
+import java.net.URLDecoder
+import java.nio.charset.StandardCharsets
+import java.util.UUID
+import javax.ws.rs.core.UriBuilder
+import kotlin.collections.set
+
+class ServiceProviderEndpointVerification(private val sendValidResponse: Boolean = true) {
+
+    private var relayState: String? = null
+    private val logoutResponse = LogoutResponseBuilder().buildObject().apply {
+        id = "a" + UUID.randomUUID().toString()
+        version = SAMLVersion.VERSION_20
+        issueInstant = DateTime()
+        issuer = IssuerBuilder().buildObject().apply { value = currentSPIssuer }
+        status = StatusBuilder().buildObject().apply {
+            StatusCodeBuilder().buildObject().apply {
+                value = if (sendValidResponse) SUCCESS else RESPONDER
+            }
+        }
+    }
+
+    /* Validations */
+
+    internal fun redirectRequestIsValid(queryParameters: Parameters): Boolean {
+        val paramMap = mutableMapOf<String, String>()
+        val request = queryParameters[SAML_REQUEST] ?: return false
+
+        paramMap[SAML_REQUEST] = request
+        queryParameters[SAML_ENCODING]?.let { paramMap[SAML_ENCODING] = it }
+        queryParameters[SIGNATURE]?.let { paramMap[SIGNATURE] = it }
+        queryParameters[SIG_ALG]?.let { paramMap[SIG_ALG] = it }
+        queryParameters[RELAY_STATE]?.let {
+            RedirectBindingVerifier().verifyRelayStateLength(it)
+            paramMap[RELAY_STATE] = it
+            relayState = it
+        }
+
+        val dom = RedirectBindingVerifier().decode(paramMap, false)
+
+        RedirectBindingVerifier().verifyNoXMLSig(dom)
+        paramMap[SIGNATURE]?.let {
+            RedirectBindingVerifier().verifyRedirectSignature(paramMap)
+            RedirectBindingVerifier().verifyRedirectDestination(dom)
+        }
+
+        CoreLogoutRequestProtocolVerifier(dom).verify()
+        return true
+    }
+
+    internal fun postRequestIsValid(body: String): Boolean {
+        val request = extractValue(body) ?: return false
+
+        relayState?.let {
+            PostFormVerifier.verifyRelayStateLength(it)
+        }
+
+        val postBindingVerifier = PostBindingVerifier()
+        val dom = postBindingVerifier.decode(request)
+        postBindingVerifier.verifyPostDestination(dom)
+
+        val requestObject = OpenSAMLUtil.fromDom(dom.ownerDocument.documentElement) as LogoutRequest
+        if (requestObject.isSigned)
+            SimpleSign().validateSignature(requestObject.signature)
+
```

Comment:
✏️ Nit: Could we do `postBindingVerifier.verifyXmlSignatures` **then** `postBindingVerifier.verifyPostDestination`? That would line up with PostBindingVerifier better.